plugins/templates/rules/undertow.j2

```diff
@@ -0,0 +1,46 @@
+{ 
+    "undertow" => {
+        {% for item in values %}
+        {% if 'configuration' in item  %}
+        "configuration" => {
+            {% for config in item['configuration'] %}
+            {% if 'filter' in config %}
+            "filter" => {
+                {% if config['filter'] =='gzip' %}
+                "gzip" => {
+                    "gzipfilter" => {}
+                }
+                {% endif %}
+            }
+            {% endif %}
+            {% endfor %}
+        },
+        {% endif %}
+        {% if 'server' in item %}
+        "server" => {
+            {% for server in item['server'] %}
+            "{{ server['name']}}" => {
+                {% if 'host' in server %}
+                "host" => {
+                    {% for host in server['host'] %}
+                    "{{ host['name'] }}" => {
+                        {% if 'filter_ref' in host %}
+                        "filter-ref" => {
+                            {% if 'gzipfilter' in host['filter_ref'][0] %}
+                            "gzipfilter" => {
+                                "predicate" => "{{ host['filter_ref'][0]['gzipfilter'][0]['predicate'] }}"
```

Comment:
you should probably assert the existence of this before starting the "for host" loop, or even the "if server in item" because the variables are set as optional

Reply:
I'm not sure I understand what you want me to change. 
On line 19 I think that I assert that server is in item. I don't assert that name is in server, because it is marked as required in the arg_spec. The same goes for host and its name.
If I have misunderstod some concept in the argument-spec, I would be happy if you could explain it to me.
I am still fairly new to the whole Ansible-model.

Reply:
If I understand the spec as it stands, both filter-ref and gzipfilter are optional; so it is possible to have a scenario like:

```
undertow:
  server:
    name: myserver
    host:
       name: myhost
       filter_ref: 
```
having `filter_ref` here an empty list.

This would generate a jcliff template like so:

```
        "server" => {
           "myserver"  => {
                "host" => {
                    "myhost"  => {
                        "filter-ref" => {}
                    }
```
I am not sure how jcliff would behave with an empty list there, did you test by any chance?

Reply:
Finally came around to testing this.
The scenario where filter-ref is an empty list seems to be valid in Jcliff. 
It just leaves the filter-ref empty.